README.md
<h1 align="center">
  PyPtax
</h1>

<h4 align="center">PyPtax is a Python library and CLI to retrieve information on <i><a href="#what-is-ptax">Ptax rates</a></i></h4>

<p align="center">
  <a href="https://pypi.python.org/pypi/pyptax">
    <img src="https://img.shields.io/pypi/v/pyptax.svg?style=flat-square" alt="Latest Version"/>
  </a>
  <a href="https://travis-ci.org/brunobcardoso/pyptax">
    <img src="https://img.shields.io/travis/brunobcardoso/pyptax/master.svg?style=flat-square" alt="Build Status"/>
  </a>
  <a href="https://codecov.io/gh/brunobcardoso/pyptax">
    <img src="https://codecov.io/gh/brunobcardoso/pyptax/branch/master/graph/badge.svg" alt="Coverage"/>
  </a>
  <a href="https://pypi.python.org/pypi/pyptax/">
    <img src="https://img.shields.io/pypi/pyversions/pyptax.svg?style=flat-square" alt="Supported Versions"/>
  </a>
  <a href='https://pyptax.readthedocs.io/en/latest/?badge=latest'>
    <img src='https://img.shields.io/readthedocs/pyptax/latest?style=flat-square' alt='Documentation Status'/>
  </a>
  <a href='https://github.com/psf/black'>
    <img src='https://img.shields.io/badge/code%20style-black-000000.svg?style=flat-square' alt='Code Style'/>
  </a>
</p>

<p align="center">
  <a href="#features">Features</a> |
  <a href="#installation">Installation</a> |
  <a href="#quickstart">Quickstart</a> |
  <a href="#documentation">Documentation</a> |
  <a href="#contributing">Contributing</a>
</p>

## What is Ptax?

[Ptax exchange rate](https://www.bcb.gov.br/conteudo/relatorioinflacao/EstudosEspeciais/EE042_A_taxa_de_cambio_de_referencia_Ptax.pdf)
is the reference exchange rate for U.S. Dollar, expressed as the amount of Brazilian Reais per one U.S. Dollar,
published by the [Central Bank of Brazil](https://www.bcb.gov.br/en).

## Features
 - Closing Ptax rates for a requested date
 - Historical Ptax rates for a requested period

## Installation
```bash
$ pip install pyptax
```

## Quickstart

### Closing Ptax rates for a requested date

**Command line:**

```bash
$ pyptax close --date 2020-01-20
```

**Module:**

```
>>> from pyptax import ptax
>>> close_report = ptax.close('2020-01-20')
>>> close_report.as_dict
{'datetime': '2020-01-20 13:09:02.871', 'bid': 4.1823, 'ask': 4.1829}
>>> close_report.datetime
'2020-01-20 13:09:02.871'
>>> close_report.bid
4.1823
>>> close_report.ask
4.1829
```

### Historical Ptax rates for a requested period

**Command line:**

```bash
$ pyptax historical --start_date 2020-01-01 --end_date 2020-01-05
```

**Module:**

```
>>> from pyptax import ptax
>>> historical_report = ptax.historical('2020-01-01', '2020-01-05')
>>> historical_report
HistoricalReport(
    start_date="2020-01-01",
    end_date="2020-01-03",
    reports=[
        CloseReport("2020-01-02 13:11:10.762", 4.0207, 4.0213),
        CloseReport("2020-01-03 13:06:22.606", 4.0516, 4.0522),
    ],
)

>>> historical_report.as_dict
{
    "start_date": "2020-01-01",
    "end_date": "2020-01-03",
    "reports": [
        {"datetime": "2020-01-02 13:11:10.762", "bid": 4.0207, "ask": 4.0213},
        {"datetime": "2020-01-03 13:06:22.606", "bid": 4.0516, "ask": 4.0522},
    ],
}
```

:::warning
:warning: Enter all dates in the format **YYYY-MM-DD**
:::

## Documentation

The full documentation is available on https://pyptax.readthedocs.io/.

## Contributing

Please see the [contributing page](https://github.com/brunobcardoso/pyptax/blob/master/CONTRIBUTING.rst)
for guidance on setting up a development environment and how to contribute.


CHANGES.rst
=============
Release Notes
=============

Version 0.5.0
=============
- Add resource to provide historical bid and ask rates for a requested time period.
- Add resource to provide intermediary rates for a requested date.
- Rename `CloseReport` to `Bulletin` and define its type on the internal attribute `bulletin_type`.
- Remove string conversion on bid and ask rates retrieved.
- Fix development installation docs.

Version 0.4.0
=============
- Add Command line interface.
- Improve exceptions handling.

  - Add ``ClientException`` to deal with response error.
  - Add ``UnavailableDataError`` to handle empty data returned.

- Add ``Close.display`` method to show data as a fixed width table for pretty printing.

Version 0.3.0
=============
- ``ptax.close`` requires date in the format YYYY-MM-DD
- Requires Python 3.7 or later
- Add documentation to readthedocs https://pyptax.readthedocs.io/.

Version 0.2.0
=============
-   ``ptax.close`` returns an instance of the ``CloseReport`` model instead of a dict.

Before:

    >>> from pyptax import ptax
    >>> ptax.close('01-20-2020')
    >>> {'datetime': '2020-01-20 13:09:02.871', 'bid': '4.1823', 'ask': '4.1829'}

After:

    >>> from pyptax import ptax
    >>> ptax.close('01-20-2020')
    >>> CloseReport(datetime='2020-01-20 13:09:02.871', bid='4.1823', ask='4.1829')
    >>> ptax.close('01-20-2020').as_dict
    >>> {'datetime': '2020-01-20 13:09:02.871', 'bid': '4.1823', 'ask': '4.1829'}

-   Create a client to deal with service request to add support for new features.
-   Add PyPI publish commands.
-   Fix typo in documentation

Version 0.1.0
=============

-   First public release.


tests/conftest.py
import pytest

from pyptax.models import Bulletin, HistoricalBulletin


@pytest.fixture
def bulletin():
    return Bulletin("2019-01-25 12:21", 4.3565, 4.3574, "close")


@pytest.fixture
def close_raw_data():
    return {
        "value": [
            {
                "dataHoraCotacao": "2020-01-16 13:04:37.351",
                "cotacaoCompra": 4.172,
                "cotacaoVenda": 4.1726,
                "bulletin_type": "close",
            }
        ]
    }


@pytest.fixture
def historical_bulletin():
    return HistoricalBulletin(
        start_date="2020-01-01",
        end_date="2020-01-03",
        bulletins=[
            Bulletin("2020-01-02 13:11:10.762", 4.0207, 4.0213, "close"),
            Bulletin("2020-01-03 13:06:22.606", 4.0516, 4.0522, "close"),
        ],
    )


@pytest.fixture
def historical_raw_data():
    return {
        "value": [
            {
                "dataHoraCotacao": "2020-01-02 13:11:10.762",
                "cotacaoCompra": 4.0207,
                "cotacaoVenda": 4.0213,
            },
            {
                "dataHoraCotacao": "2020-01-03 13:06:22.606",
                "cotacaoCompra": 4.0516,
                "cotacaoVenda": 4.0522,
            },
        ]
    }


@pytest.fixture
def intermediary_raw_data():
    return {
        "value": [
            {
                "cotacaoCompra": 4.0101,
                "cotacaoVenda": 4.0107,
                "dataHoraCotacao": "2020-01-02 10:08:18.114",
                "tipoBoletim": "Abertura",
            },
            {
                "cotacaoCompra": 4.0118,
                "cotacaoVenda": 4.0124,
                "dataHoraCotacao": "2020-01-02 11:03:40.704",
                "tipoBoletim": "Intermediário",
            },
            {
                "cotacaoCompra": 4.0302,
                "cotacaoVenda": 4.0308,
                "dataHoraCotacao": "2020-01-02 12:10:55.168",
                "tipoBoletim": "Intermediário",
            },
            {
                "cotacaoCompra": 4.0305,
                "cotacaoVenda": 4.0311,
                "dataHoraCotacao": "2020-01-02 13:11:10.756",
                "tipoBoletim": "Intermediário",
            },
            {
                "cotacaoCompra": 4.0207,
                "cotacaoVenda": 4.0213,
                "dataHoraCotacao": "2020-01-02 13:11:10.762",
                "tipoBoletim": "Fechamento PTAX",
            },
        ]
    }


tests/test_cli.py
from unittest import mock

import pytest
from click.testing import CliRunner

import pyptax
from pyptax.cli import cli
from pyptax.exceptions import ClientError


@pytest.fixture
def cli_runner():
    return CliRunner()


def test_version(cli_runner):
    result = cli_runner.invoke(cli, ["--version"])

    assert pyptax.__version__ in result.output
    assert result.exit_code == 0


def test_no_subcommand_shows_help_message(cli_runner):
    result = cli_runner.invoke(cli)

    assert "PyPtax Command Line Interface" in result.output
    assert result.exit_code == 0


@mock.patch("pyptax.ptax.close", autospec=True)
def test_close(mock_close, cli_runner):
    date = "2020-01-20"

    result = cli_runner.invoke(cli, ["close", "--date", date])

    mock_close.assert_called_once_with(date)
    assert result.exit_code == 0


@mock.patch("pyptax.ptax.close", autospec=True)
def test_close_error(mock_close, cli_runner):
    date = "2020-01-20"
    expected = "Test Error"
    mock_close.side_effect = ClientError(expected)

    result = cli_runner.invoke(cli, ["close", "--date", date])

    mock_close.assert_called_once_with(date)
    assert expected in result.output
    assert result.exit_code == 0


@mock.patch("pyptax.ptax.historical", autospec=True)
def test_historical(mock_historical, cli_runner):
    start_date = "2020-01-01"
    end_date = "2020-01-03"

    result = cli_runner.invoke(
        cli, ["historical", "--start_date", start_date, "--end_date", end_date]
    )

    mock_historical.assert_called_once_with(start_date, end_date)
    assert result.exit_code == 0


@mock.patch("pyptax.ptax.historical", autospec=True)
def test_historical_error(mock_historical, cli_runner):
    start_date = "2020-01-01"
    end_date = "2020-01-03"
    expected = "Test Error"
    mock_historical.side_effect = ClientError(expected)

    result = cli_runner.invoke(
        cli, ["historical", "--start_date", start_date, "--end_date", end_date]
    )

    mock_historical.assert_called_once_with(start_date, end_date)
    assert expected in result.output
    assert result.exit_code == 0


@mock.patch("pyptax.ptax.intermediary", autospec=True)
def test_intermediary(mock_intermediary, cli_runner):
    date = "2020-01-02"

    result = cli_runner.invoke(cli, ["intermediary", "--date", date])

    mock_intermediary.assert_called_once_with(date)
    assert result.exit_code == 0


@mock.patch("pyptax.ptax.intermediary", autospec=True)
def test_intermediary_error(mock_intermediary, cli_runner):
    date = "2020-01-01"
    expected = "Test Error"
    mock_intermediary.side_effect = ClientError(expected)

    result = cli_runner.invoke(cli, ["intermediary", "--date", date])

    mock_intermediary.assert_called_once_with(date)
    assert expected in result.output
    assert result.exit_code == 0


tests/test_ptax.py
import responses

from pyptax import ptax, settings
from pyptax.models import Bulletin, HistoricalBulletin, IntermediaryBulletin


@responses.activate
def test_close(close_raw_data):
    date = "2020-01-16"

    responses.add(
        responses.GET,
        f"{settings.SERVICE_URL}{settings.CLOSE_RESOURCE}?@dataCotacao='01-16-2020'&$format=json",
        json=close_raw_data,
    )

    close = ptax.close(date)

    assert len(responses.calls) == 1
    assert isinstance(close, Bulletin)


@responses.activate
def test_historical(historical_raw_data):
    params = "@dataInicial='01-01-2020'&@dataFinalCotacao='01-03-2020'&$format=json"
    responses.add(
        responses.GET,
        f"{settings.SERVICE_URL}{settings.HISTORICAL_RESOURCE}?{params}",
        json=historical_raw_data,
    )

    historical = ptax.historical("2020-01-01", "2020-01-03")

    assert len(responses.calls) == 1
    assert isinstance(historical, HistoricalBulletin)


@responses.activate
def test_intermediary(intermediary_raw_data):
    params = "@moeda='USD'&@dataCotacao='01-02-2020'&$format=json"
    responses.add(
        responses.GET,
        f"{settings.SERVICE_URL}{settings.INTERMEDIARY_RESOURCE}?{params}",
        json=intermediary_raw_data,
    )

    intermediary = ptax.intermediary("2020-01-02")

    assert len(responses.calls) == 1
    assert isinstance(intermediary, IntermediaryBulletin)


pyptax/cli.py
import click

import pyptax
from pyptax.exceptions import ClientError
from pyptax.exceptions import DateFormatError
from pyptax.exceptions import UnavailableDataError

LOGO_CLI = (
    " /$$$$$$$            /$$$$$$$    /$$\n"
    "| $$__  $$          | $$__  $$  | $$\n"
    "| $$  \\ $$ /$$   /$$| $$  \\ $$ /$$$$$$    /$$$$$$  /$$   /$$\n"
    "| $$$$$$$/| $$  | $$| $$$$$$$/|_  $$_/   |____  $$|  $$ /$$/\n"
    "| $$____/ | $$  | $$| $$____/   | $$      /$$$$$$$ \\  $$$$/\n"
    "| $$      | $$  | $$| $$        | $$ /$$ /$$__  $$  >$$  $$\n"
    "| $$      |  $$$$$$$| $$        |  $$$$/|  $$$$$$$ /$$/\\  $$\n"
    "|__/       \\____  $$|__/         \\___/   \\_______/|__/  \\__/\n"
    "           /$$  | $$\n"
    "          |  $$$$$$/\n"
    "           \\______/\n"
)


@click.group(invoke_without_command=True)
@click.option("--version", "-v", is_flag=True, default=False)
@click.pass_context
def cli(ctx, version):
    """
    PyPtax Command Line Interface.

    PyPtax is a Python library to retrieve information on Ptax rates.
    """
    if not ctx.invoked_subcommand and version:
        click.secho(pyptax.__version__)
        ctx.exit()

    elif not ctx.invoked_subcommand:
        click.secho(LOGO_CLI, fg="green")
        click.secho(ctx.get_help(), fg="green")
        ctx.exit()


@cli.command("close")
@click.option("--date", "-d", required=True, type=str)
def close(date):
    """Provide bid and ask rates for the requested date."""
    try:
        click.secho(str(pyptax.ptax.close(date).display()), fg="green")
    except (ClientError, DateFormatError, UnavailableDataError) as exc:
        click.secho(str(exc), fg="red")


@cli.command("historical")
@click.option("--start_date", "-sd", required=True, type=str)
@click.option("--end_date", "-ed", required=True, type=str)
def historical(start_date, end_date):
    """Provide bid and ask rates for the requested time period."""
    try:
        click.secho(
            str(pyptax.ptax.historical(start_date, end_date).display()), fg="green"
        )
    except (ClientError, DateFormatError, UnavailableDataError) as exc:
        click.secho(str(exc), fg="red")


pyptax/models.py
from dataclasses import asdict
from dataclasses import dataclass
from typing import List
from typing import Optional

from tabulate import tabulate


@dataclass
class CloseReport:
    datetime: str
    bid: float
    ask: float

    @property
    def as_dict(self):
        return asdict(self)

    def display(self, fmt: Optional[str] = "psql") -> str:
        return tabulate(self.as_dict.items(), tablefmt=fmt)


@dataclass
class HistoricalReport:
    start_date: str
    end_date: str
    reports: List[CloseReport]

    @property
    def as_dict(self):
        return asdict(self)

    def display(self, fmt: Optional[str] = "psql") -> str:
        rows = ((report.datetime, report.bid, report.ask) for report in self.reports)
        return tabulate(rows, headers=self.reports[0].as_dict.keys(), tablefmt=fmt)


pyptax/resources.py
from .helpers import DateParser
from .models import CloseReport
from .models import HistoricalReport
from pyptax import settings
from pyptax.exceptions import UnavailableDataError


class CloseResource:
    path = settings.CLOSE_RESOURCE

    def __init__(self, date):
        self.date = date
        self.parsed_date = DateParser(date).parse()

    @property
    def params(self):
        return f"@dataCotacao={self.parsed_date!r}&$format=json"

    def parse(self, raw_data):
        try:
            data = raw_data["value"][0]
        except IndexError:
            raise UnavailableDataError(
                f"Unavailable rates for the requested date:{self.date!r}"
            )
        datetime = data["dataHoraCotacao"]
        bid = data["cotacaoCompra"]
        ask = data["cotacaoVenda"]
        return CloseReport(datetime, bid, ask)


class HistoricalResource:
    path = settings.HISTORICAL_RESOURCE

    def __init__(self, start_date, end_date):
        self.start_date = start_date
        self.end_date = end_date
        self.parsed_start = DateParser(start_date).parse()
        self.parsed_end = DateParser(end_date).parse()

    @property
    def params(self):
        return f"@dataInicial={self.parsed_start!r}&@dataFinalCotacao={self.parsed_end!r}&$format=json"

    def parse(self, raw_data):
        reports_list = raw_data["value"]
        if not reports_list:
            raise UnavailableDataError(
                f"Unavailable rates for the requested range: {self.start_date!r} - {self.end_date!r}"
            )

        close_reports = [
            CloseReport(
                report["dataHoraCotacao"],
                report["cotacaoCompra"],
                report["cotacaoVenda"],
            )
            for report in reports_list
        ]

        return HistoricalReport(self.start_date, self.end_date, close_reports)


pyptax/ptax.py
from .client import Ptax
from .models import CloseReport
from .models import HistoricalReport
from .resources import CloseResource
from .resources import HistoricalResource


def close(date: str) -> CloseReport:
    """
    Retrieve closing Ptax rates on a certain date.

    Parameters
    ----------
    date
        Year, month and day of the date to be searched. Format - "YYYY-MM-DD"

    Returns
    -------
    CloseReport
        A CloseReport object with datetime, bid and ask attributes

    Raises
    ------
    DateFormatError
        If fails to parse the informed date

    Examples
    --------
    >>> close_report = close("2020-01-20")
    >>> close_report
    CloseReport(datetime=2020-01-20 13:09:02.871, bid=4.1823, ask=4.1829)
    >>> close_report.bid
    4.1823
    >>> close_report.ask
    4.1829
    >>> close_report.as_dict
    {'datetime': '2020-01-20 13:09:02.871', 'bid': 4.1823, 'ask': 4.1829}

    """
    resource = CloseResource(date)
    return Ptax(resource).response()


def historical(start_date: str, end_date: str) -> HistoricalReport:
    """
    Retrieve historical closing Ptax rates for the requested time period.

    Parameters
    ----------
    start_date
        Beginning of the time period to be searched. Format - 'YYYY-MM-DD'
    end_date
        End of the time period to be searched. Format - 'YYYY-MM-DD'

    Returns
    -------
    HistoricalReport
        A HistoricalReport object with start_date, end_date and reports with a list
        of CloseReports.

    Raises
    ------
    DateFormatError
        If fails to parse the informed date
    ClientError
        If Ptax Service response returns an error
    UnavailableDataError
        If receives an empty list from Ptax Service

    Examples
    --------
    >>> historical_report = historical("2020-01-02", "2020-01-03")
    >>> historical_report
    HistoricalReport(
        start_date="2020-01-02",
        end_date="2020-01-04",
        reports=[
            CloseReport(datetime="2020-01-02 13:11:10.762", bid=4.0207, ask=4.0213),
            CloseReport(datetime="2020-01-03 13:06:22.606", bid=4.0516, ask=4.0522),
        ],
    )
    >>> historical_report.as_dict
    {
        "start_date": "2020-01-02",
        "end_date": "2020-01-04",
        "reports": [
            {"datetime": "2020-01-02 13:11:10.762", "bid": 4.0207, "ask": 4.0213},
            {"datetime": "2020-01-03 13:06:22.606", "bid": 4.0516, "ask": 4.0522},
        ],
    }
    """
    resource = HistoricalResource(start_date, end_date)
    return Ptax(resource).response()
